src/aobasis/ao_util.F

```diff
@@ -190,57 +190,59 @@ FUNCTION exp_radius(l, alpha, threshold, prefactor, epsin) RESULT(radius)
          a = ABS(alpha)
       END IF
 
-      IF (threshold == 0.0_dp) THEN
-         CPABORT("The requested threshold is zero")
-      ELSE
+      IF (threshold .NE. 0.0_dp) THEN
          t = ABS(threshold)
+      ELSE
+         CPABORT("The requested threshold is zero")
       END IF
 
-      IF (prefactor == 0.0_dp) THEN
-         RETURN
-      ELSE
+      IF (prefactor .NE. 0.0_dp) THEN
          d = ABS(prefactor)
+      ELSE
+         RETURN
       END IF
 
       ! Calculate the Maximum g(r)
-      r = SQRT(0.5_dp*REAL(l, dp)/a)
-      ar2 = a*r*r
+      r0 = SQRT(0.5_dp*REAL(l, dp)/a)
+      ar2 = a*r0*r0
 
       IF (l == 0) THEN
          g = d
       ELSE
-         g = d*r**l*EXP(-ar2)
+         g = d*r0**l*EXP(-ar2)
       END IF
 
-      IF (t > g) THEN
+      IF (g .LT. t) THEN
          RETURN
       END IF
 
-      rlow = r
-      rhigh = 2.0_dp*rlow+1.0_dp
-      iter = 0
-      DO
-         iter = iter+1
-         IF (iter .GT. maxiter) THEN
-            CPABORT("Maximum number of iterations exceeded")
-         END IF
-         g = d*rhigh**l*EXP(-a*rhigh**2)
-         IF (g < t) EXIT
-         rlow = rhigh
-         rhigh = 2.0_dp*rlow+1.0_dp
+      r1 = scale*r0+1.0_dp
+      DO iter = 1, maxiter
+         ar2 = a*r1*r1
+         g = d*r1**l*EXP(-ar2)
+         IF (g .LT. t) EXIT
+         r0 = r1
+         r1 = scale*r0+1.0_dp
       ENDDO
+      ! avoid system calls in tight/arithmetic loops
+      IF (iter .EQ. maxiter) THEN
```

Comment:
Shouldn't this rather be: `IF (iter .EQ. maxiter+1)`?

Reply:
```console
$ cat iter.f90 
program main
    integer :: idx

    do idx = 1, 10
        print *, idx
    enddo

    print *, "final idx: ", idx
end program
$ ./iter 
           1
           2
           3
           4
           5
           6
           7
           8
           9
          10
 final idx:           11
```

Reply:
It's a draft ;-)

... and I think there is a much bigger issue. Let me describe:

* The first iteration captures an interval (r0, r1] such that g(r1) < threshold.
* The second iteration "bisects" the interval such that rd=r1-r0 and rd < epsiter.
* The function is supposed to yield a radius r in (r0, r1] such that r->min.

The original code was supposed to be a bisection i.e., the interval contracts with 0.5. So, I think Golden Section Search (GSS) should work as well and give an egde over bisection (contraction != 0.5). In any case (Bisection or GSS),  unimodal functions are the prerequisite/limitation for finding an extreme value (minimum in this case). The unimodal property is likely ensured by the code that produces the initial interval. The captured function g(r) over (r0, r1] looks like coming from the top and strongly decaying to zero ("function plot").

Now the problem: all tests (e.g., tests/LIBTEST/test_eri_mme_accuracy_longrange.inp) seem to be very sensitive i.e., even if a contraction produces an r that is slightly smaller and even holds g(r) < threshold, it seems the tests will fail. The test calculates reference values that are "theoretical upper bounds" and then claims that (some) test values exceeded the "theoretical upper bound".

I think I will play a bit further with this...

Reply:
Indeed, I messed up the final iteration index. Anyhow, that is not a problem at this point and can be fixed later.

Reply:
Btw, I am (intentionally) lumping together the iterations from the first part and the latter part:

```fortran
DO iter = iter, maxiter
```

I think what matters are the number of function-evaluations g(r). Btw, in most cases the algorithm does ~10 iterations according to the new counting (as just explained).

Reply:
Is there any particular reason for optimizing this specific routine?

> Now the problem: all tests (e.g., tests/LIBTEST/test_eri_mme_accuracy_longrange.inp) seem to be very sensitive i.e., even if a contraction produces an r that is slightly smaller and even holds g(r) < threshold, it seems the tests will fail. The test calculates reference values that are "theoretical upper bounds" and then claims that (some) test values exceeded the "theoretical upper bound".

If it is just the `eri_mme` tests, I'd need to have a look. I'm aware that `test_eri_mme_accuracy_longrange.inp` is not very stable and even the claim about the theoretical upper bound is not strictly true, so you can ignore that test.

FYI: I wrote a faster implementation of `exp_radius` myself for the `eri_mme` integrals since `exp_radius` was a bottleneck, you can find it [here](https://github.com/cp2k/cp2k/blob/master/src/eri_mme/eri_mme_gaussian.F#L262-L292). Instead of bisection, it directly iterates over equidistant points, but it avoids evaluating exponentials in the loop. However I would not trust that implementation to be stable for general use.

Reply:
Thank you, I will take a look. Let's call this exercise "without real need" so far. I just came accross this routine since it's all scalar. I only wanted to remove the system-call from the loop with EXP in it (and I will only do that if everything else fails). Anyhow, I am attracted by the fact how instable it makes CP2K although changing the routine from "bisection" to "GSS" should only not make a difference for the matter if exp_radius. However, what called "theoretical upper bound" might be only a set of number calibrated to the current error margin.

Thank you again for sharing your work!